fix(docs-infra): apply various a11y aio improvements

dario-piotrowicz opened the pull request:
changes:
 - currently the navigation `ul` used in `aio-top-menu` has a role of
   navigation, but listitems should be owned by list parents
   (see more: https://www.w3.org/TR/wai-aria-1.1/#listitem)
   so wrap the `ul` in a `nav` and remove the `role="navigation"` from the
   `ul` element to fix such issue
 - wrap the main aio mat-toolbar in a header element to provide better
   accessibility
 - remove reduntant role="main"
 - (try to) make sure that there in only a single main in the aio page
 - (try to) make sure all elements are included in landmarks
 - (try to) make sure the main navs have different labels

resolves #16938
resolves #44562

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

jessicajaniuk reviewed (approved):
LGTM :cookie:

I love seeing this a11y work, @dario-piotrowicz. Keep up the great work!

dario-piotrowicz commented:
@gkalpak could you have a quick look at this one please? :blush:

gkalpak commented on aio/content/guide/hierarchical-dependency-injection.md:
```diff
@@ -191,8 +191,6 @@ directive.
 Components and directives on the same element share an injector.
 
 
-{@a resolution-rules}
```
TBH, I am not a huge fan of removing these (esp. since they are actually used in links), because it makes the links to these sections fragile. See [Section header anchors][1] for more details on why.

But I can't think of a good solution. If you don't have any ideas either, then I am Ok dropping them, I guess 😞 

[1]: https://angular.io/guide/docs-style-guide#section-header-anchors

gkalpak replied:
Actually, one solution could be to change the `@a` anchors to something slightly different and update existing links. This way, old links will work (for now) thanks to auto-generated IDs for the `<h>` elements and new URLs will work thanks to the `@a` anchors.

Alternatively, we could consider changing `@a` to generate `<a name="...">` instead of `<a id="...">` (if they behave the same).

I am not thrilled about any of these solutions either, so leaving it up to you, @dario-piotrowicz :smiley:

dario-piotrowicz replied:
to be totally honest I removed those `{@a`s but I thought this change should not be in this PR as they are all over the place and I am just removing a handful which I just happened to come across, I figured I could keep the change in as it still looked like an improvement (but it seems like it is not without drawbacks)

What I would prefer to do maybe is to just revert these removals here and take the conversation somewhere else (I can open an issue and we can look into that), and then I can do a follow up PR in which I can try to remove (or whatever we decide to do) all of them instead of just those I came across

What do you think? would that work for you? :slightly_smiling_face: 

(by the way, what would you think of a programmatic solution, as in, in the app, when a page is loaded we can check, and if there are multiple elements with the same id we remove the anchor/empty one, does that sound too hacky? :thinking:)

dario-piotrowicz replied:
I've removed it from the current PR :slightly_smiling_face::+1:

gkalpak commented on aio/src/styles/2-modules/code/_code-theme.scss:
```diff
@@ -66,10 +66,10 @@
       .copy-button {
-        color: constants.$blue-grey-200;
+        color: constants.$blue-grey-500;
 
         &:hover {
-          color: constants.$mediumgray;
+          color: constants.$deepgray;
```
These new values don't seem to work well when using the dark theme.
Can you please check all updates styles on both the light and dark theme (both visually and for a11y contrast requirements)? :pray:

dario-piotrowicz replied:
mh.... I don't remember if I checked both themes :sweat: 

sorry I'll have a look :slightly_smiling_face: :bow:

dario-piotrowicz replied:
yeah I definitely didn't check the dark theme! :sweat:

dario-piotrowicz replied:
amended :slightly_smiling_face: sorry for the silly mistake! :sweat:

gkalpak commented on aio/src/styles/0-base/_typography-theme.scss:
```diff
@@ -43,7 +43,7 @@
 
   .sidenav-content a {
-    color: if($is-dark-theme, constants.$lightblue, constants.$blue);
+    color: if($is-dark-theme, constants.$lightblue, constants.$darkblue);
```
While this increases the contrast between the anchor text and the background, it makes it harder to distinguish the anchor text from the regular text, which is not good a11y/UX either. Could we maybe make this a little lighter (and still meet contrast requirements)?

dario-piotrowicz replied:
I'll have a look :slightly_smiling_face::+1:

dario-piotrowicz replied:
I've updated it to `#055cff` (please let me know if you mind the hex there instead of a constant :sweat:)

This is how it looked before:
![Screenshot at 2022-03-16 22-11-53](https://user-images.githubusercontent.com/61631103/158700014-1bfc682f-7df3-4f72-92cd-6915b5b05d0b.png)
and this is after:
![Screenshot at 2022-03-16 22-11-34](https://user-images.githubusercontent.com/61631103/158700037-45b30b5b-216d-4c6f-a1c3-68d102a244d8.png)

we do get the AA check

what do you think? :slightly_smiling_face: 
is the blue a bit too vibrant? :sweat:

gkalpak replied:
It is a bit too vibrant for my taste. It feels a lit out of place with the test of our color-palette 😞
(We could really use a designer 👀 😛)

gkalpak reviewed (changes requested):
Great work, @dario-piotrowicz 👏

BTW, in the future, it is a good idea to break up such changes into separate commits :wink:

dario-piotrowicz commented:
@gkalpak thanks a lot for the speedy review! :heart: :heart_eyes: :heart: :heart_eyes: :heart: :heart_eyes: :heart: :heart_eyes:

dario-piotrowicz commented:
> Great work, @dario-piotrowicz clap
> 
> BTW, in the future, it is a good idea to break up such changes into separate commits wink

About the commits, yeah sorry I did started that way but then things were intertwining and I just figured I could just mash it into a single one :sweat_smile:, as part of my next changes to this PR I can put the effort and try to divide stuff in ways that make sense, I am totally fine with that, it should not be too much of an issue if you want :slightly_smiling_face: 

(also the commits would be a bit small, hopefully you don't mind that :smile:)

gkalpak commented:
> as part of my next changes to this PR I can put the effort and try to divide stuff in ways that make sense

No, no need for that. Just something to keep in mind for future PRs :wink:

> also the commits would be a bit small, hopefully you don't mind that 😄

We certainly don't mind small commits (as long as they are self-contained) :smiley:

dario-piotrowicz committed:
fix(docs-infra): remove navigation role to aio-top-menu ul

currently the navigation ul used in aio-top-menu has a role of
navigation, but listitems should be owned by list parents
(see more: https://www.w3.org/TR/wai-aria-1.1/#listitem)
so wrap the ul in a nav and remove the role="navigation" from the
ul element to fix such issue

resolves #44562
resolves #16938 (the second point)

dario-piotrowicz committed:
fix(docs-infra): wrap the main aio mat-toolbar in a header

wrap the main aio mat-toolbar in a header element to provide better
accessibility

resolves #16938 (the first point)

dario-piotrowicz committed:
fix(docs-infra): remove redundant main roles

remove redundant main role as pages should always have a
single element with a main role (also remove the role assigne
to the main tag as that is implied)

dario-piotrowicz committed:
fix(docs-infra): assign different aria labels to main aio navs

assign different aria labels to the primary nav and the one used for
guides and docs, so that impaired users can more easily distinguish the
two

dario-piotrowicz committed:
fix(docs-infra): update a11y min-scores

update (decrease) the value of some of the accessibility MIN_SCORES_PER_PAGE
after recent changes

dario-piotrowicz commented:
@gkalpak I've addressed your comments, please have another look :blush:

mary-poppins commented:
You can preview 718f709 at https://pr45209-718f709.ngbuilds.io/.

gkalpak commented on aio/src/styles/0-base/_typography-theme.scss:
```diff
@@ -43,7 +43,7 @@
 
   .sidenav-content a {
-    color: if($is-dark-theme, constants.$lightblue, constants.$blue);
+    color: if($is-dark-theme, constants.$lightblue, #055cff);
```
Note: If we decide to keep this (or any color we end up with), we need to make sure that links inside code-blocks have the same color (and the color has to be accessible with both the default white-ish background and the gray code-block background).

mary-poppins commented:
You can preview 7409b7e at https://pr45209-7409b7e.ngbuilds.io/.

dario-piotrowicz commented on aio/scripts/test-aio-a11y.mjs:
```diff
@@ -29,8 +29,8 @@ const MIN_SCORES_PER_PAGE = {
   'guide/docs-style-guide': 96,
-  'start/start-routing': 98,
-  'tutorial': 98,
+  'start/start-routing': 97,
+  'tutorial': 97,
```
@gkalpak instead of increasing the scores have decreases without the color contrast improvements :sweat: , I'd imagine it could be that my changes are somehow exposing more of the view to lighthouse which finds more issues (possibly color contrast ones). Because otherwise I would not imagine that my changes to improve a11y are actually making it worse :cry: 

What do you think? :sweat: 

(PS: remember that with the color contrast improvements all of these were at 100 :thinking: , at the very least we can should be back to all 100s when we fix the color contrast issues :thinking:)

gkalpak replied:
I checked with Lighthouse and it turns out that the lowered score is a result of less checks running (so the same ones that fail now account for a larger percentage). In particular, some `[role]`-related checks ([this][1] and [that][2]) are now skipped.

So, even if the score here is lowered, a11y didn't really got worse 🤷‍♂️ 

[1]: https://web.dev/aria-required-attr
[2]: https://web.dev/aria-roles

dario-piotrowicz replied:
ah I see :thinking: 

I didn't really get that, thanks so much for looking into it :smile: :heart: 

As I said anyways, let's hope that we can fix the color contrast things soon and go back to 100s :sunglasses:

gkalpak replied:
🤞

dario-piotrowicz commented:
@gkalpak I've removed the color changes as those seems to be a bit tricky and could be handled separately, please have another look :slightly_smiling_face: 

(also now unfortunately the a11y scores are now going slightly down instead of going up :sob:)

gkalpak reviewed (approved):
:shipit:

dylhunn commented:
This PR was merged into the repository by commit 16e635ca492a31856f344b53c0649fd3edc179c3.
